File: android-catalogue/catalogue/src/main/java/xyz/ksharma/prisma/catalogue/playground/CodeBlock.kt
package xyz.ksharma.prisma.catalogue.playground

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.LocalClipboardManager
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.unit.dp
import kotlinx.coroutines.delay
import xyz.ksharma.prisma.components.icons.PrismaIcons
import xyz.ksharma.prisma.coreui.themed
import xyz.ksharma.prisma.tokens.PrismaRadius
import xyz.ksharma.prisma.tokens.PrismaSemanticColors
import xyz.ksharma.prisma.tokens.PrismaSpacing
import xyz.ksharma.prisma.tokens.PrismaTypography

/**
 * Live, copyable code snippet showing how to use the component with the
 * current knob values. Sits below the playground's States gallery.
 */
@Composable
public fun CodeBlock(
    code: String,
    modifier: Modifier = Modifier,
    language: String = "kotlin",
) {
    val clipboard = LocalClipboardManager.current
    var copied by remember { mutableStateOf(false) }
    LaunchedEffect(copied) {
        if (copied) {
            delay(1500)
            copied = false
        }
    }

    Column(
        modifier = modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(PrismaRadius.Md))
            .background(PrismaSemanticColors.SurfaceSunken.themed())
            .border(
                1.dp,
                PrismaSemanticColors.BorderSubtle.themed(),
                RoundedCornerShape(PrismaRadius.Md),
            ),
    ) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = PrismaSpacing.Sp4, vertical = PrismaSpacing.Sp2),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.SpaceBetween,
        ) {
            Text(
                text = language.uppercase(),
                style = PrismaTypography.LabelSm,
                color = PrismaSemanticColors.TextTertiary.themed(),
            )
            Row(
                modifier = Modifier
                    .clip(RoundedCornerShape(PrismaRadius.Sm))
                    .clickable(
                        role = Role.Button,
                        onClick = {
                            clipboard.setText(AnnotatedString(code))
                            copied = true
                        },
                    )
                    .padding(horizontal = PrismaSpacing.Sp2, vertical = PrismaSpacing.Sp1),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(PrismaSpacing.Sp1),
            ) {
                Icon(
                    painter = painterResource(if (copied) PrismaIcons.Check else PrismaIcons.Copy),
                    contentDescription = if (copied) "Copied" else "Copy",
                    tint = if (copied) PrismaSemanticColors.StatusSuccessDefault.themed()
                           else PrismaSemanticColors.TextSecondary.themed(),
                    modifier = Modifier.size(14.dp),
                )
                Text(
                    text = if (copied) "Copied" else "Copy",
                    style = PrismaTypography.LabelSm,
                    color = if (copied) PrismaSemanticColors.StatusSuccessDefault.themed()
                            else PrismaSemanticColors.TextSecondary.themed(),
                )
            }
        }
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .background(PrismaSemanticColors.SurfaceBase.themed())
                .horizontalScroll(rememberScrollState())
                .padding(PrismaSpacing.Sp4),
        ) {
            Text(
                text = code,
                style = PrismaTypography.BodySm.copy(fontFamily = FontFamily.Monospace),
                color = PrismaSemanticColors.TextPrimary.themed(),
            )
        }
    }
}

File: ios-catalogue/App/Playground/CodeBlock.swift
import SwiftUI
import CoreUI
#if canImport(UIKit)
import UIKit
#endif

/// Live, copyable code snippet showing how to use the component with the
/// current knob values. Sits below the playground's States gallery.
struct CodeBlock: View {
    @Environment(\.colorScheme) private var scheme
    let code: String
    var language: String = "swift"

    @State private var copied: Bool = false

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            HStack {
                Text(language.uppercased())
                    .font(PrismaTypography.labelSm.font)
                    .foregroundStyle(PrismaSemanticColors.textTertiary.themed(scheme))
                Spacer()
                Button(action: copy) {
                    HStack(spacing: PrismaSpacing.sp1) {
                        Image(prisma: copied ? .check : .copy)
                            .renderingMode(.template)
                            .resizable()
                            .frame(width: 14, height: 14)
                            .foregroundStyle(
                                copied
                                    ? PrismaSemanticColors.statusSuccessDefault.themed(scheme)
                                    : PrismaSemanticColors.textSecondary.themed(scheme)
                            )
                        Text(copied ? "Copied" : "Copy")
                            .font(PrismaTypography.labelSm.font)
                            .foregroundStyle(
                                copied
                                    ? PrismaSemanticColors.statusSuccessDefault.themed(scheme)
                                    : PrismaSemanticColors.textSecondary.themed(scheme)
                            )
                    }
                    .padding(.horizontal, PrismaSpacing.sp2)
                    .padding(.vertical, PrismaSpacing.sp1)
                }
                .buttonStyle(.plain)
            }
            .padding(.horizontal, PrismaSpacing.sp4)
            .padding(.vertical, PrismaSpacing.sp2)

            ScrollView(.horizontal, showsIndicators: false) {
                Text(code)
                    .font(.system(.footnote, design: .monospaced))
                    .foregroundStyle(PrismaSemanticColors.textPrimary.themed(scheme))
                    .padding(PrismaSpacing.sp4)
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(PrismaSemanticColors.surfaceBase.themed(scheme))
        }
        .background(PrismaSemanticColors.surfaceSunken.themed(scheme))
        .clipShape(RoundedRectangle(cornerRadius: PrismaRadius.md))
        .overlay(
            RoundedRectangle(cornerRadius: PrismaRadius.md)
                .strokeBorder(PrismaSemanticColors.borderSubtle.themed(scheme), lineWidth: 1)
        )
    }

    private func copy() {
        #if canImport(UIKit)
        UIPasteboard.general.string = code
        #endif
        copied = true
        DispatchQueue.main.asyncAfter(deadline: .now() + 1.5) { copied = false }
    }
}
